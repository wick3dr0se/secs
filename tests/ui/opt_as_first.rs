use secs::World;

fn optional_components() {
    let world = World::default();

    world.spawn((1_u32,));
    world.spawn((10_u32, "foo"));
    let mut results = vec![];
    world.query(|_, i: &u32, s: Option<&&str>| results.push((*i, s.map(|s| *s))));
    results.sort();
    assert_eq!(&results[..], &[(1, None), (10, Some("foo"))]);

    let mut results = vec![];
    world.query(|_, s: Option<&&str>, i: &u32| results.push((*i, s.map(|s| *s))));
    //~^ ERROR: is not a valid query
    results.sort();
    assert_eq!(&results[..], &[(10, Some("foo"))]);
}
