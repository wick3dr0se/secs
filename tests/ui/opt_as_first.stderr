error[E0277]: `{closure@tests/ui/opt_as_first.rs:14:17: 14:47}` is not a valid query
   --> tests/ui/opt_as_first.rs:14:17
    |
14  |     world.query(|_, s: Option<&&str>, i: &u32| results.push((*i, s.map(|s| *s))));
    |           ----- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |           |
    |           required by a bound introduced by this call
    |
    = help: the trait `secs::query::Query<_>` is not implemented for closure `{closure@tests/ui/opt_as_first.rs:14:17: 14:47}`
    = note: only tuples with 1 or up to 5 elements can be used as queries
note: required by a bound in `World::query`
   --> $DIR/src/world.rs
    |
LLL |     pub fn query<Q: Query<T>, T>(&self, f: Q) {
    |                     ^^^^^^^^ required by this bound in `World::query`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
